src/sentry/event_manager.py

```diff
@@ -907,6 +913,75 @@ def get_culprit(data):
     )
 
 
+def _find_group_id(all_hashes):
+    for h in all_hashes:
+        if h.group_id is not None:
+            return h.group_id
+        if h.group_tombstone_id is not None:
+            raise HashDiscarded("Matches group tombstone %s" % h.group_tombstone_id)
+
+    return None
+
+
+def _save_aggregate2(event, hashes, release, **kwargs):
+    project = event.project
+
+    all_hashes = [GroupHash.objects.get_or_create(project=project, hash=hash)[0] for hash in hashes]
+    existing_group_id = _find_group_id(all_hashes)
+
+    if existing_group_id is None:
+        with transaction.atomic():
+            all_hashes = list(
+                GroupHash.objects.filter(id__in=[h.id for h in all_hashes]).select_for_update()
+            )
+
+            existing_group_id = _find_group_id(all_hashes)
+
+            if existing_group_id is None:
+                short_id = project.next_short_id()
+
+                # it's possible the release was deleted between
+                # when we queried for the release and now, so
+                # make sure it still exists
+                first_release = kwargs.pop("first_release", None)
+
+                group = Group.objects.create(
+                    project=project,
+                    short_id=short_id,
+                    first_release_id=Release.objects.filter(id=first_release.id)
+                    .values_list("id", flat=True)
+                    .first()
+                    if first_release
+                    else None,
+                    **kwargs,
+                )
+
+                # invariant: existing_group_id is None, therefore all hashes
+                # have group_id=None, therefore none of them can be locked in
+                # migration either
+                GroupHash.objects.filter(id__in=[h.id for h in all_hashes]).update(group=group)
```

Comment:
Old code excluded all LOCKED_IN_MIGRATION hashes here. This is now gone because not having races means we can conclude none of our hashes are locked in migration.